### ISSUE TITLE
REAL Locking a CPU thread?

### ISSUE THREAD

**Issue by Luadih** [miniant-git/REAL#9]
**REAL version**: v0.2.0

**Describe the bug**
When REAL is open in the background or is open at the startup and then closed, it "locks" up a thread in the cpu, by now, i think its only a visual lock, but if it has a performance decrease it is a slight one at least in the benchmark i tested.

**To Reproduce**
1. Launch REAL
2. Try to get the CPU Usage to max.
3. Watch the CPU Usage in Task Manager or HWMonitor
4. 1 Thread does not reach its max. usage

**Expected behavior**
CPU Usage monitoring showing what it should (100% CPU Usage)

**Screenshots**
https://prnt.sc/tsjr57

**Comment by bthanl** [comment]
Can confirm that this IS an issue. Performance decrease is definitely there. I've tested multiple times using various benchmarks and they all have similar results. My CPU is an i5 9600k which has 6 cores and the benchmark scores/results are around 5/6 of what they should be when REAL is not open.

It doesn't seem to be "locking" the core, just hogging it. Setting the affinity of a program to only use the core that is "locked" will result in that core having normal CPU usage, albeit programs obviously will run terribly.

**Comment by skipfie** [comment]
I can confirm this issue as well, getting low to no usage on core #9 on a 2700x benchmarking on cinebench with REAL running. Closing the program and restarting the PC "frees" up the core, and improves the benchmark by ~200 pts.

I've also noticed windows not sleeping due to the audio stream being used

**Comment by fdrva** [comment]
Same with 8600k. I get about 20% performance cut in all applications. Benchmarks, baking and games.
It's not REAL issue but AudioClient3 thing. I don't think real dev can fix this.

**Comment by Devocub** [comment]
it's windows audio service + scheduler to blame. It does it to be able to maintain low latency.
Good thing is to add notification about this in the app and balloon if it's trayed.
https://forums.tomshardware.com/threads/1-core-not-working-on-amd-ryzen-3-3100.3763694/

**Comment by xiiyuki** [comment]
Has anyone ever found a fix for that?? I was literally using REAL for 1,5 years trying to find the culprit of why my CPU 0 doesn't randomly work

**Comment by Devocub** [comment]
@Oxi-nya
Restart Audio service.bat 
```
@echo off
net stop Audiosrv /y
net start Audiosrv /y
```

**Comment by xiiyuki** [comment]
> @Oxi-nya Restart Audio service.bat
> 
> ```
> @echo off
> net stop Audiosrv /y
> net start Audiosrv /y
> ```

After restarting it will revert the buffer size, right?? considering that I'd restart `Audiosrv` while the program is running

**Comment by Luadih** [comment]
I don't use REAL anymore but something like this should be possible to use to launch REAL, loop till it detects REAL is closed, and then restart the audio service right? It should just need a few line changes I think

https://www.reddit.com/r/Batch/comments/po07ql/comment/hcychjw/?utm_source=share&utm_medium=web2x&context=3

**Comment by Devocub** [comment]
@Oxi-nya restarting audio service is like restarting pc but only audio is restarted. So things are at default if not rerequested by some program (for example osu reinitializes audio engine, if it had low latency thing it would reask system to make buffer small again. REAL doesn't reinitialize).

@Luadih the solution you provided is bad and messy. Don't use it. It can be added into REAL properly. This cpu core locking is windows bug - it should be fixed in first place.

**Comment by Luadih** [comment]
@Devocub Yeah, just meant for it to be a band-aid for people who want to use it and want to work around it using any way possible. REAL and Windows should include a fix for it. Though I don't know if REAL will get updated now, unless someone forks it and works on it.

EDIT: Changed like 3 words for a better meaning

**Comment by LastExceed** [comment]
The behaviour seems to be by design. From point 3 in https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio#audio-stack-improvements :

> When an application uses buffer sizes below a certain threshold to render and capture audio, Windows enters a special mode, where it manages its resources in a way that avoids interference between the audio streaming and other subsystems. This will reduce the interruptions in the execution of the audio subsystem and minimize the probability of audio glitches.

Processing audio in such small intervals makes it extremely susceptible to interrupts, especially since glitches from even the tiniest underruns can be immediately audible. So the scheduler tries to keep 1 cpu thread available at all times, to be used by the audio subsystem at a moments notice.

But the thread isn't entirely reserved, its basically just last in line whenever a cpu thread gets assigned a workload by the scheduler. As discussed in the LAL thread (linked above), it can still be utilized for other things when they are explicitly limited to that specific thread, or when all other threads are busy. So the performance penalty boils down to
1. the increased demand of the sound server,
2. the reduced thermal balance across cpu cores,

and is imo nothing to worry about.

As for the persistence after closing REAL: I unfortunately can't reproduce this. Maybe it really was a bug and it already got patched, or maybe another audio stream from a different application was still running, keeping the engine in the high performance mode. But either way, its not something to fix in REAL, so I'd say this issue can be closed as "by design".

---
PS: for the osu! players reading this, the new "experimental audio mode" option in Lazer obsoletes REAL / LAL (its built-in)

**Comment by Luadih** [comment]
Thanks to having made this issue, I was updated and learned about a lot of things lol

Sadly I cannot test if the persistence after closing REAL still happens and say it got fixed by a windows update or something by discarding my PC of the equation, since I upgraded my CPU and other things in the 5 years since I opened the issue.

What I can say is, having tried LAL (LowAudioLatency) quite recently on my current Ryzen 5 5600 in Windows 10 21H2 I checked if any cores got locked and it didn't seem to happen, but I also checked only a single day and didn't go too in depth with the testing. Some people were reporting a similar issue with LAL so I wouldn't completely discard the issue though.

Edit: I did test it the same way by using Cinebench R15 and checking the core utilization shown by task manager.

Edit2: A lot of people -> Some people
